### DISCUSSION TITLE
"docker stop nominatim" does not trigger "service postgresql stop"

### DISCUSSION THREAD

**Discussion by JonasHeidelberg** [mediagis/nominatim-docker#284] [Q&A]
Whenever I have to stop my nomatim-docker container and restart it, I am told by postgres (output of `tail -f /var/log/postgresql/postgresql-12-main.log`): 
`database system was not properly shut down; automatic recovery in progress`

The /app/start.sh script contains a block

```
stopServices() {
  service apache2 stop
  service postgresql stop
}
trap stopServices TERM
```

which looks to me like it should trigger  `service postgresql stop` whenever I do `docker stop nominatim` on the host machine. However, this does not work for me. Only if I manually `docker exec -it nominatim service postgresql stop` before `docker stop nominatim` I can avoid the postgresql error message about improper shutdown. 

**Update:** I tried increasing the time between docker sending SIGTERM and SIGKILL from 10 to 30 seconds (`docker stop -t 30 nominatim`) but this had no effect. The container still terminated with errorcode 137, which I understand means the PID 1 process did not end gracefully, and I still see the postgresql error message when launching the container the next time.

In case this matters, I roughly follow the 2nd example from https://hub.docker.com/r/mediagis/nominatim/ - more specifically
```
docker run -it --rm --shm-size=1g \
  -e PBF_PATH=/localpath-to-my-region.osm.pbf \
  -e IMPORT_WIKIPEDIA=false \
  -e NOMINATIM_PASSWORD=very_secure_password \
  -v nominatim-data:/var/lib/postgresql/12/main \
  -p 8080:8080 \
  --name nominatim \
  mediagis/nominatim:3.7
```

Am I doing something wrong?

**Comment by leonardehrenfried** [comment]
I think it's just a bug in the image.

I've seen it before and it's quite worrying however postgres seems to recover from the cold shutdown.

If you have an idea about how to shutdown cleanly, we'd love to have a PR.

**Reply by pgassmann** [reply]
The reason is that in Dockerfile, the CMD is set as a string, therefore it is launched in /bin/sh which runs the program in a subshell that does not receive the signals.

fixed in #373
